File: src/main/java/clone/bzm/lounge/example/adapter/in/web/ExampleController.java
package clone.bzm.lounge.example.adapter.in.web;

import clone.bzm.lounge.example.application.port.in.ExampleLoadUseCase;
import clone.bzm.lounge.example.application.port.in.ExampleSaveUseCase;
import clone.bzm.lounge.example.domain.Example;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.util.UriComponentsBuilder;

import java.net.URI;
import java.util.List;

@RequiredArgsConstructor
@RestController
class ExampleController {

    private final ExampleLoadUseCase exampleLoadUseCase;
    private final ExampleSaveUseCase exampleSaveUseCase;

    @GetMapping("/api/example/{exampleId}")
    ResponseEntity<ExampleResponse> findExample(@PathVariable Long exampleId) {
        Example selected = exampleLoadUseCase.findExample(exampleId);

        return ResponseEntity.ok(ExampleResponse.of(selected));
    }

    @GetMapping("/api/example")
    ResponseEntity<List<ExampleResponse>> findExample() {
        List<Example> selected = exampleLoadUseCase.findExample();

        return ResponseEntity.ok(ExampleResponse.of(selected));
    }

    @PostMapping("/api/example")
    ResponseEntity<ExampleResponse> createExample(@RequestBody @Valid ExampleCreateRequest request,
                                                  UriComponentsBuilder uriComponentsBuilder) {

        Example created = exampleSaveUseCase.saveExample(request.getName());

        URI location = uriComponentsBuilder.path("/api/example/{exampleId}")
                .buildAndExpand(created.getId())
                .toUri();

        return ResponseEntity.created(location)
                .body(ExampleResponse.of(created));
    }

    @DeleteMapping("/api/example/{exampleId}")
    ResponseEntity<ExampleResponse> deleteExample(@PathVariable Long exampleId) {
        Example deleted = exampleSaveUseCase.deleteExample(exampleId);

        return ResponseEntity.ok(ExampleResponse.of(deleted));
    }

    @PutMapping("/api/example/{exampleId}")
    ResponseEntity<ExampleResponse> updateExample(@PathVariable Long exampleId,
                                                  @RequestBody ExampleUpdateRequest request) {

        Example updated = exampleSaveUseCase.updateExample(exampleId, request.getName());

        return ResponseEntity.ok(ExampleResponse.of(updated));
    }
}

File: src/main/kotlin/clone/bzm/lounge/example/adapter/in/web/ExampleController.kt
package clone.bzm.lounge.example.adapter.`in`.web

import clone.bzm.lounge.example.adapter.`in`.web.dto.ExampleCreateRequest
import clone.bzm.lounge.example.application.port.`in`.CreateExampleUseCase
import clone.bzm.lounge.example.application.port.`in`.LoadExampleUseCase
import clone.bzm.lounge.example.domain.Example
import org.springframework.web.bind.annotation.*

@RestController
class ExampleController(
    private val exampleUseCase: LoadExampleUseCase,
    private val createExampleUseCase: CreateExampleUseCase) {

    // todo: 외부로 나가는건 domain이 나가는게 아니고 별도 ExampleResponse로 변경 필요한듯

    @GetMapping("/api/example/{exampleId}")
    fun findExampleById(@PathVariable exampleId: String): Example {
        return exampleUseCase.findExampleById(exampleId);
    }

    @PostMapping("/api/example")
    fun createExample(@RequestBody exampleCreateRequest: ExampleCreateRequest): Example {
        return createExampleUseCase.createExample(exampleCreateRequest)
    }
}